### ISSUE TITLE
ntopng may hang on shutdown

### ISSUE THREAD

**Issue by cavedon** [ntop/ntopng#1424]
In case ntopng is listening on an interface with no traffic, after receiving a SIGTERM ntopng will hang forever in `pthread_join(pollLoop, &res)`.
The problem is that ntopng tries to interrupt the sniffer thread with `pcap_breakloop(pcap_handle)`.
However this will not interrupt `pcap_next_ex()`, on which the sniffer thread is blocked on. Interruption will happen only of there is traffic on the interface.
I tried to do some research and it looks like there is a way to get `pcap_next_ex()` interrupted. The solution may be to use select/epoll of the pcap file descriptor, so that the blocking call is now select/epoll, and it can be interrupted by writing onto a pipe or more easily via a periodic timeout.

Verified on Debian stretch on both 2.4 and latest dev branch.

**Comment by lucaderi** [comment]
@cavedon I have a PC with a wlan0 interface and no traffic. Termination occurs properly as pcap_next_ex() will return 0 in case there is no traffic according to the timeout set in pcap_open_live()

Can you please tell me how to reproduce the issue as I am apparently unable to do it?

**Comment by lucaderi** [comment]
Closed for inactivity, will reopen if necessary. In the meantime a few changes have been made and hopefully the problem is gone.

**Comment by tobiasdam** [comment]
I think I am experiencing the issue described above.

I am using `Raspbian Stretch` on a Raspberry Pi 3 and installed the `ntopng v.3.2.171203` stable version using the apt-stable repository. 

The system is acting as a Wi-Fi access point, therefore ntopng is listening on `eth0` and additionally on `wlan0`. In case no device is connected to the Wi-Fi, stopping the ntopng service hangs after receiving `SIGTERM` until systemd kills the service with `SIGKILL` after the default timeout of 90 seconds. 

After some debugging, I noticed following things:
- Stopping the service right after starting the service (< ca. 10s) works without issues
- Stopping the service finishes immediately after a devices connects to the Wi-Fi

**Comment by emanuele-f** [comment]
Can you attach a gdb debugger to ntopng while it is blocked `gdb --pid ntopng_pid` and print a `thread apply all bt`?

**Comment by tobiasdam** [comment]
Of course:

```gdb
(gdb) thread apply all bt                                                                                                                                                                                   [55/388]

Thread 12 (Thread 0x6830f320 (LWP 3707)):
#0  0x7633ed50 in poll () at ../sysdeps/unix/syscall-template.S:84
#1  0x76efa8f4 in ?? () from /usr/lib/arm-linux-gnueabihf/libpcap.so.0.8
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 11 (Thread 0x6ab0f320 (LWP 3702)):
#0  0x7694094c in __pthread_cond_wait (cond=0x3bb1668, mutex=0x3bb1928) at pthread_cond_wait.c:186
#1  0x000e53cc in Mutex::cond_wait (this=0x3bb1928, condvar=0x3bb1668) at /home/deri/ntopng/include/Mutex.h:49
#2  0x000e51b4 in ThreadPool::dequeueJob (this=0x3bb1660, waitIfEmpty=true) at src/ThreadPool.cpp:134
#3  0x000e4f70 in ThreadPool::run (this=0x3bb1660) at src/ThreadPool.cpp:81
#4  0x000e4cd0 in doRun (ptr=0x3bb1660) at src/ThreadPool.cpp:29
#5  0x76939fc4 in start_thread (arg=0x6ab0f320) at pthread_create.c:335
Backtrace stopped: Cannot access memory at address 0x400006f8

Thread 10 (Thread 0x6bb0f320 (LWP 3700)):
#0  0x7694094c in __pthread_cond_wait (cond=0x3bb11d8, mutex=0x3bb1498) at pthread_cond_wait.c:186
#1  0x000e53cc in Mutex::cond_wait (this=0x3bb1498, condvar=0x3bb11d8) at /home/deri/ntopng/include/Mutex.h:49
#2  0x000e51b4 in ThreadPool::dequeueJob (this=0x3bb11d0, waitIfEmpty=true) at src/ThreadPool.cpp:134
#3  0x000e4f70 in ThreadPool::run (this=0x3bb11d0) at src/ThreadPool.cpp:81
#4  0x000e4cd0 in doRun (ptr=0x3bb11d0) at src/ThreadPool.cpp:29
#5  0x76939fc4 in start_thread (arg=0x6bb0f320) at pthread_create.c:335
Backtrace stopped: Cannot access memory at address 0x400006f8

Thread 9 (Thread 0x6cb0f320 (LWP 3698)):
#0  0x7694094c in __pthread_cond_wait (cond=0x3bb0d48, mutex=0x3bb1008) at pthread_cond_wait.c:186
#1  0x000e53cc in Mutex::cond_wait (this=0x3bb1008, condvar=0x3bb0d48) at /home/deri/ntopng/include/Mutex.h:49
#2  0x000e51b4 in ThreadPool::dequeueJob (this=0x3bb0d40, waitIfEmpty=true) at src/ThreadPool.cpp:134
#3  0x000e4f70 in ThreadPool::run (this=0x3bb0d40) at src/ThreadPool.cpp:81
#4  0x000e4cd0 in doRun (ptr=0x3bb0d40) at src/ThreadPool.cpp:29
#5  0x76939fc4 in start_thread (arg=0x6cb0f320) at pthread_create.c:335
Backtrace stopped: Cannot access memory at address 0x400006f8

Thread 8 (Thread 0x6db0f320 (LWP 3696)):
#0  0x7694094c in __pthread_cond_wait (cond=0x3bb08d8, mutex=0x349aad0) at pthread_cond_wait.c:186
#1  0x000e53cc in Mutex::cond_wait (this=0x349aad0, condvar=0x3bb08d8) at /home/deri/ntopng/include/Mutex.h:49
#2  0x000e51b4 in ThreadPool::dequeueJob (this=0x3bb08d0, waitIfEmpty=true) at src/ThreadPool.cpp:134
#3  0x000e4f70 in ThreadPool::run (this=0x3bb08d0) at src/ThreadPool.cpp:81
#4  0x000e4cd0 in doRun (ptr=0x3bb08d0) at src/ThreadPool.cpp:29
#5  0x76939fc4 in start_thread (arg=0x6db0f320) at pthread_create.c:335
Backtrace stopped: Cannot access memory at address 0x400006f8

Thread 7 (Thread 0x6f8ff320 (LWP 3690)):                                                                                                                                                                            
#0  0x7694094c in __pthread_cond_wait (cond=0x24825c0, mutex=0x2482070) at pthread_cond_wait.c:186                                                                                                                  
#1  0x00092c5c in consume_socket (ctx=0x2481fd0, sp=0x6ed006a4) at src/../third-party/mongoose/mongoose.c:5094                                                                                                      
#2  0x00092e40 in worker_thread (thread_func_param=0x2481fd0) at src/../third-party/mongoose/mongoose.c:5132
#3  0x76939fc4 in start_thread (arg=0x6f8ff320) at pthread_create.c:335
---Type <return> to continue, or q <return> to quit---
#4  0x76347c68 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:76 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 6 (Thread 0x700ff320 (LWP 3689)):
#0  0x7694094c in __pthread_cond_wait (cond=0x24825c0, mutex=0x2482070) at pthread_cond_wait.c:186
#1  0x00092c5c in consume_socket (ctx=0x2481fd0, sp=0x6f0006a4) at src/../third-party/mongoose/mongoose.c:5094
#2  0x00092e40 in worker_thread (thread_func_param=0x2481fd0) at src/../third-party/mongoose/mongoose.c:5132
#3  0x76939fc4 in start_thread (arg=0x700ff320) at pthread_create.c:335
#4  0x76347c68 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:76 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 5 (Thread 0x708ff320 (LWP 3688)):
#0  0x7694094c in __pthread_cond_wait (cond=0x24825c0, mutex=0x2482070) at pthread_cond_wait.c:186
#1  0x00092c5c in consume_socket (ctx=0x2481fd0, sp=0x6ef006a4) at src/../third-party/mongoose/mongoose.c:5094
#2  0x00092e40 in worker_thread (thread_func_param=0x2481fd0) at src/../third-party/mongoose/mongoose.c:5132
#3  0x76939fc4 in start_thread (arg=0x708ff320) at pthread_create.c:335
#4  0x76347c68 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:76 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 4 (Thread 0x714cf320 (LWP 3687)):
#0  0x7694094c in __pthread_cond_wait (cond=0x24825c0, mutex=0x2482070) at pthread_cond_wait.c:186
#1  0x00092c5c in consume_socket (ctx=0x2481fd0, sp=0x70a006a4) at src/../third-party/mongoose/mongoose.c:5094
#2  0x00092e40 in worker_thread (thread_func_param=0x2481fd0) at src/../third-party/mongoose/mongoose.c:5132
#3  0x76939fc4 in start_thread (arg=0x714cf320) at pthread_create.c:335
#4  0x76347c68 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:76 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 3 (Thread 0x71ccf320 (LWP 3686)):
#0  0x7694094c in __pthread_cond_wait (cond=0x24825c0, mutex=0x2482070) at pthread_cond_wait.c:186
#1  0x00092c5c in consume_socket (ctx=0x2481fd0, sp=0x709006a4) at src/../third-party/mongoose/mongoose.c:5094
#2  0x00092e40 in worker_thread (thread_func_param=0x2481fd0) at src/../third-party/mongoose/mongoose.c:5132
#3  0x76939fc4 in start_thread (arg=0x71ccf320) at pthread_create.c:335
#4  0x76347c68 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:76 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 2 (Thread 0x724cf320 (LWP 3685)):             
#0  0x7633ed50 in poll () at ../sysdeps/unix/syscall-template.S:84                                        
#1  0x000934b8 in master_thread (thread_func_param=0x2481fd0) at src/../third-party/mongoose/mongoose.c:5273
#2  0x76939fc4 in start_thread (arg=0x724cf320) at pthread_create.c:335                                   
#3  0x76347c68 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:76 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)                                

Thread 1 (Thread 0x73105070 (LWP 3662)):             
#0  0x7693b410 in pthread_join (threadid=<optimized out>, thread_return=0x7e805654) at pthread_join.c:90  
#1  0x000c0218 in PcapInterface::shutdown (this=0x1558570) at src/PcapInterface.cpp:232                   
#2  0x000d6224 in Ntop::shutdown (this=0xedf5a0) at src/Ntop.cpp:1658                                     
#3  0x000c0884 in sigproc (sig=15) at src/main.cpp:59                                                     
---Type <return> to continue, or q <return> to quit---                                                    
#4  <signal handler called>                          
#5  0x76313054 in nanosleep () at ../sysdeps/unix/syscall-template.S:84                                   
#6  0x76341b0c in usleep (useconds=<optimized out>) at ../sysdeps/posix/usleep.c:32                       
#7  0x000d2b14 in Ntop::start (this=0xedf5a0) at src/Ntop.cpp:453                                         
#8  0x000c17b8 in main (argc=2, argv=0x7e805e44) at src/main.cpp:402                                      
(gdb)                                                
```

**Comment by emanuele-f** [comment]
Fixed with f84adf37fb50493600881f279ecea82c0fcaa669
